jlm/rvsdg/node.hpp

```diff
@@ -846,6 +846,40 @@ class node
   std::vector<std::unique_ptr<node_output>> outputs_;
 };
 
+class node_input_iterator final : public input::iterator<input>
+{
+public:
+  constexpr explicit node_input_iterator(input * value)
+      : input::iterator<input>(value)
+  {}
+
+  [[nodiscard]] input *
+  next() const override
+  {
+    auto index = value()->index() + 1;
+    auto node = input::GetNode(*value());
+
+    return index >= node->ninputs() ? nullptr : node->input(index);
+  }
+};
+
+class node_const_input_iterator final : public input::iterator<const input>
+{
+public:
+  constexpr explicit node_const_input_iterator(const input * value)
+      : input::iterator<const input>(value)
+  {}
+
+  [[nodiscard]] const input *
+  next() const override
+  {
+    auto index = value()->index() + 1;
+    auto node = input::GetNode(*value());
+
+    return index >= node->ninputs() ? nullptr : node->input(index);
```

Comment:
Is node guaranteed to be not NULL here?

Reply:
removed the entire helper class -- I think working with std::vector<...> of things is really a lot easier and less constraining on the internal data structures. I would be happy to revisit later (e.g. expose std::span things).